Core/HttpHandlerProxy.cs

```diff
@@ -0,0 +1,151 @@
+using System.Collections.Generic;
+using System.Net.Http.Headers;
+using System.Net.Http;
+using System.Net;
+using System.Security.Authentication;
+using System.Text;
+using Microsoft.Net.Http.Headers;
+using System.Threading.Tasks;
+using System.Threading;
+using Microsoft.Exchange.WebServices.Data;
+using Microsoft.Exchange.WebServices.NETStandard.Extensions;
+
+namespace Microsoft.Exchange.WebServices.NETStandard.Core
+{
+    internal class HttpHandlerProxy : DelegatingHandler
+    {
+        class SimpleHttpMessageHandlerProxy : DelegatingHandler
+        {
+            public SimpleHttpMessageHandlerProxy(HttpMessageHandler messageHandler)
+                : base(messageHandler) 
+            {
+            }
+
+            internal Task<HttpResponseMessage> SendImplAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                return base.SendAsync(request, cancellationToken);
+            }
+        }
+
+        private readonly string httpClientName;
+        private readonly HttpMessageHandler httpMessageHandler;
+        private readonly HttpClientHandler httpClientHandler;
+
+        /// <summary>
+        /// Gets or sets the cookie container.
+        /// </summary>
+        /// <value>The cookie container.</value>
+        public CookieContainer CookieContainer { get; set; }
+
+        /// <summary>
+        /// Gets or sets authentication information for the request.
+        /// </summary>
+        /// <returns>An <see cref="T:System.Net.ICredentials"/> that contains the authentication credentials associated with the request. The default is null.</returns>
+        public ICredentials Credentials { get; set; }
+
+        /// <summary>
+        /// Gets or sets proxy information for the request.
+        /// </summary>
+        public IWebProxy Proxy { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value that indicates whether to send an authenticate header with the request.
+        /// </summary>
+        /// <returns>true to send a WWW-authenticate HTTP header with requests after authentication has taken place; otherwise, false. The default is false.</returns>
+        public bool PreAuthenticate { get; set; }
+
+        /// <summary>
+        /// Gets or sets a <see cref="T:System.Boolean"/> value that controls whether default credentials are sent with requests.
+        /// </summary>
+        /// <returns>true if the default credentials are used; otherwise false. The default value is false.</returns>
+        public bool UseDefaultCredentials { get; set; }
+
+        public HttpHandlerProxy(IHttpMessageHandlerFactory httpMessageHandlerFactory, string httpClientName)
+        {
+            this.httpClientName = httpClientName;
+            this.httpMessageHandler = httpMessageHandlerFactory?.CreateHandler(httpClientName);
+            this.httpClientHandler = null == httpMessageHandlerFactory ? HttpClientExtensions.CreateHttpClientHandler(httpClientName) : null;

```

Comment:
1. In which cases httpMessageHandlerFactory may be null?
2. names of 2 handlers are unclear. Can you rename them? e.g. sharedHttpMessageHandler and dedicatedHttpMessageHandler. I'd even change this to single InnerHandler and bool initialized=true/false, bool shared/privateHandler?

Reply:
1. If client use library in old way without DI
2. rename is ok, but set them both to InnerHandler is not possible because in case of negotiate authentication mode I need make first call through shared handler and if response is 401 with www-authenticate header then create dedicated handler and set credentials to it, also once request was made for InnerHandler you can change it. Take a look at HttpHandlerProxy.SendAsync imlementation

Reply:
1 ok
2 ok, agree on innerhandler (can't change it once set). still clarity in vars will make it easier.